### DISCUSSION TITLE
Postgresql Views - Improve Nested Relationship Performance

### DISCUSSION THREAD

**Discussion by RegisWeduLabs** [directus/directus#12521] [Feature Requests]
Hi @directus team,

I think this is more of a question/discussion, so I'm not sure this is the best media to discuss this, but I'll take my chance.

When fetching an item with nested relations, depending on the depth you need, you end up with a "loop" where the parent document ends up nested in the children.

To illustrate my "issue", if, for example, you have the following structure:

root => Translations => O2M => O2M => Other, and the "fields" parameters is set to `"*.*.*.*.*"`, you end up with the following:

root => Translations => root_Id => Translations => root_Id => Translations => root_id...   And it goes on and on if you request deeper nested object.

As we requested the root to begin with, I cannot see a use case where you would want the root object to be re-nested in children object, multiple time.  If you have an object with a depth of 8-10, you end up with a stack limit exceeded, or the instance becomes unresponsive - or it throws a 504 error.

It's probably not a "typical" use case, but generally speaking, if we prevent this behaviour, we get more performance out of the queries in general, and we end up being able to use more complex structures without running into this issue.

We worked around it by using the other options (getting the fields/relationships then building the object using multiple queries etc...), bit I just thought it would make sense generally speaking. 

I'm willing to have a look at this and create a PR - but before I do so, I wanted to make sure it is something you would consider.  Otherwise, I think I'll just end up building a custom module for these scenarios to extend Directus.

I would appreciate your thoughts on the matter.

Regards!

**Comment by rijkvanzanten** [comment]
Ideally, you should only fetch the data you actually need, rather than relying on Directus to fetch everything on every level. This is also outlined [in the docs](https://docs.directus.io/reference/query/#fields): 

> **Performance & Size**
> While the fields wildcard is very useful for debugging purposes, we recommend only requesting specific fields for production use. By only requesting the fields you really need, you can speed up the request, and reduce the overall output size.

This is also one of the benefits of GraphQL, as it inherently requires you to do that.

**Reply by RegisWeduLabs** [reply]
Our use case for this particular project is that we are using it as a "headless" web page builder...  So we need the full object (in our case, it's the pages) - and because webpage structure has a "logical" structure where you end up with Divs in divs, etc...  We get into this issue.  I also still think it would most likely help with performance overall for nested relationships.  So I understand the "main" idea where we want it as light as possible, but in those use case, the object is quite big and it is expected...  What we do is we fetch the full object and then we dynamically build the webpage based on the object.  It actually works quite well, but we need to "work" around.

**Comment by RegisWeduLabs** [comment]
I was reading trough the meaningful part of #12476 , and it seems somehow related.

I went trough the code today (where the runAST happens), and it would be a bit difficult to change it the way it is at the moment...  I was able to prevent the loop in the object itself but not in the recursive function calls (yet).

Thinking out loud, would it be a good idea to implement views on selected tables?  Example, in the table configuration, add an option that would create a view based on the schema - therefore offloading the workload to the database server?  Knex supports it - so it should be something quite straightforward to do...  It would probably be good for directus core tables too - it'd most likely speed up the app generally speaking.

Let me know your thoughts?
Thanks!

**Reply by rijkvanzanten** [reply]
Views are definitely something we're interested in supporting (also see https://github.com/directus/directus/discussions/3097 regarding that). Now that knex supports it, it should be way more doable to start supporting them in Directus as well 👍🏻

**Reply by RegisWeduLabs** [reply]
All right - makes sense.  I am working on an extension for postgres that includes creating/querying views, and a simple checkbox to create a view from the table settings.  It will give a proof of concept, and I will share the repo here once I have something that works.

**Reply by rijkvanzanten** [reply]
That's awesome!

**Reply by RegisWeduLabs** [reply]
@rijkvanzanten, hey, I have done some pretty good progress to build postgresql views based on directus schemas/relations. It’s close. I’m working on the last bit for m2a in m2a. I should have something in the next few days. Just wanted to keep you posted.

**Comment by RegisWeduLabs** [comment]
Hey @rijkvanzanten,

Sorry, it took a bit longer than expected.  Here is the link to the extension: https://github.com/Hubelia/directus-postgresql-json-views
It is still very much a functional proof of concept and still need quite a bit of work, but I wanted to share ASAP so you can see where I was going with this.  By all means, if you see any better way to do this, let me know.  It works quite well for our use case, (used to take over 2 minutes to get our big structure of data, now below 2 seconds for the same object) and I think it could potentially help in many other use cases too.  Let me know your thoughts?

Thanks and best regards,

**Reply by rijkvanzanten** [reply]
This is awesome! Definitely keeping an eye on where you'll take it next 🙂

**Reply by RegisWeduLabs** [reply]
Thanks, still have a bit of work to do but I intend to focus on it for the next few days so we have an initial release ASAP.  Will keep you posted.

**Comment by RegisWeduLabs** [comment]
Hey @rijkvanzanten, I've added a few features to the Extension - materialized views, revamped a few things and fixed a few bugs.

Still need to implement a few things, and I wanted to check what is the best option for authentication...  It's becoming a bit more dependant to directus and I want it to stay in line with good practices.  It's the next logical step as it's not secured.  I've seen a few examples for authentication in some other extensions but nothing ideal from my point of view...

Thanks!

**Reply by RegisWeduLabs** [reply]
Awright, based on: https://github.com/directus/directus/discussions/4551 - I am able to make sure the user is authenticated - AND filter the collections included in the view based on the permission.  So all good!

**Comment by bamboowonder** [comment]
I'm excited to see where you take this.   I'd love to be making use of views, material or non for many of our collections in directus.

**Reply by RegisWeduLabs** [reply]
I'll get back to it in the next few days to implement authentication and queries, at this point I'd be more comfortable for people to use it in production.  It actually works really well for us (but our use case is that we need all the data).  I've already implemented materialized views and automate the view update via the Directus webhook (for the time being).  It's amazingly fast - as the view contains the JSON data - so no manipulation needed.  I'd really like to hear from another user - then again, not production ready for obvious reasons.  We can continue discussions for this extension here https://github.com/Hubelia/directus-postgresql-json-views/discussions - it'll be easier to track.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
